### DISCUSSION TITLE
How to integrade and develope third-party packages/extensions with `starlette-admin`?

### DISCUSSION THREAD

**Discussion by hasansezertasan** [jowilf/starlette-admin#478] [Q&A] [answered]
Hello everyone 👋 I have started to develop a package called [hasansezertasan/starlette-admin-fields](https://github.com/hasansezertasan/starlette-admin-fields) but there is one thing that bothers me. How can I properly integrate it with `starlette-admin`? Emphasize the "properly".

I do recommend checking #439 before rereading further.

In that issue, we decided to go with a third-party package for additional fields, and in this https://github.com/jowilf/starlette-admin/issues/439#issuecomment-1870580099 I asked about how to integrate a third-party package with `starlette-admin` but there is still not an answer. It's blurry. What is the proper way?

Flask documentation has a page about it: [Flask Extension Development — Flask Documentation (3.0.x)](https://flask.palletsprojects.com/en/3.0.x/extensiondev/) and it makes it clear to develop an extension for the Flask ecosystem.

For now, none of the fields has any additional HTML template but some of them require some static CSS and JS files.

So how to extend `statics` endpoint with a `statics` folder from a third-party package also same question goes for `templates` too. Maybe we can add an interface to `starlette-admin` for easier integration or maybe we can add an extension development guide page in the documentation like Flask.

And also, should we move `TinyMCEEditorField` to that package or leave it here?

**Comment by hasansezertasan** [comment]
Here are some resources, discussions, examples, StackOverflow questions, etc that I found on the internet.

- Static Files:
  - encode/starlette#537
  - encode/starlette#625
  - encode/starlette#1350
  - [Static Files - Starlette](https://www.starlette.io/staticfiles/)
- Templates
  - encode/starlette#1214
  - [Templates - Starlette](https://www.starlette.io/templates/)
  - [Templating - Starlette Core](https://accent-starlette.github.io/starlette-core/templating/)
- Flask Ecosystem
  - [python - How to use a Jinja2 extension direct from a Flask render_template - Stack Overflow](https://stackoverflow.com/questions/52683386/)
  - [helloflask/bootstrap-flask](https://github.com/helloflask/bootstrap-flask) quite similar to what I want to build.
  - [helloflask/flask-ckeditor: CKEditor integration for Flask, including image upload, code syntax highlight, and more.](https://github.com/helloflask/flask-ckeditor)
  - [digital-land/govuk-jinja-components](https://github.com/digital-land/govuk-jinja-components): This one handled it pretty well...
  - [LandRegistry/govuk-frontend-jinja: GOV.UK Frontend Jinja Macros](https://github.com/LandRegistry/govuk-frontend-jinja)
  - [LandRegistry/govuk-frontend-wtf: GOV.UK Frontend WTForms Widgets](https://github.com/LandRegistry/govuk-frontend-wtf)
  - [nathancahill/Flask-Edits: Editable Content in Flask](https://github.com/nathancahill/flask-edits)
  - [artemShelest/flask-admin-subview: Embedded Flask-Admin list views](https://github.com/artemShelest/flask-admin-subview)
  - [cllen/flask-admin-tablefield: tablefield for flask-admin!](https://github.com/cllen/flask-admin-tablefield)

**Comment by hasansezertasan** [comment]
WDYT about this one @jowilf?

I'm hyped about the `starlette-admin-fields` package. I believe something like mentioned in this https://github.com/jowilf/starlette-admin/issues/439#issuecomment-1870580099 could solve our problem.

**Reply by jowilf** [reply]
it might be easiest to provide a function, like `init_admin`, that takes the admin as an argument and adds any additional configuration needed for the package to work. For instance, an extra loader or a mounted static route could be added.

To add an extra loader, for example, you could use the following code:
```py
admin.templates.env.loader.append(PackageLoader(""))
```

Please note that this is only a suggestion and I haven't looked into it deeply.

**Reply by hasansezertasan** [reply]
> it might be easiest to provide a function, like `install...`, that takes the admin as an argument and adds any additional configuration needed for the package to work. For instance, an extra loader or a mounted static route could be added.

So you are recommending adding a function to `starlette-admin-fields` package that extends the templates loader? If yes, I thought about that but how are we going to extend `StaticFiles`?

Maybe something like this:

```python
def install_fields(admin: Any) -> None:
    admin.templates.env.loader.append(PackageLoader("starlette_admin_fields", "templates"))
    # What about static files?
```

I also didn't try this but I know `StaticFile` object doesn't have something similar, it doesn't allow extending the statics folder on the go. You have to pass `packages=["starlette_admin", "starlette_admin_fields"]` when initializing.

**Reply by jowilf** [reply]
```py
admin.routes.add(Mount('unique_name', app=StaticFiles(...))
```

**Reply by hasansezertasan** [reply]
> ```python
> admin.routes.add(Mount('unique_name', app=StaticFiles(...))
> ```

Well, that's an implementation. I'll give it a try and present you a demo.

Edit: One question, what should I name this unique name? `statics` reserved by `starlette-admin`. Does `statics-saf` (`saf` => `Starlette-Admin-Fields`) sound good?

**Comment by hasansezertasan** [comment, marked as answer]
You can find the full example at [starlette-admin-fields](https://github.com/hasansezertasan/starlette-admin-fields) repo.

@jowilf Thank you for your help.
